<a href="https://colab.research.google.com/github/melodyhe21/Predicting-Potability/blob/main/Melody's_Copy_of_3Data_Clean_Environment_Async3_CM1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Gaining Insight Through Computational Modeling - Async 3
## Data Cleaning: Environment

#### Author - Rachel Calder
#### Student - Melody He
#### Description - Data visualization is the use of graphs, charts, or other visualizations to explore the data and identify patterns and trends. Through this notebook, we will explore data and create some visualizations to help us understands its patterns and trends. **Save a copy of this notebook to your student folder.**

##### *This notebook is written in python computer language. To run a block of code, click on the block, and press the play button.

The data used in this notebook is provided by Kaggle. The data was collcted to assess the potability of water. You can learn more about the dataset here: https://www.kaggle.com/adityakadiwal/water-potability



## Loading Python Libraries


In [ ]:
## Load necessary libraries

#pandas helps us handle data tables
#importing pandas as pd means we are giving pandas a short nickname (pd) that is quicker to type
import pandas as pd

#numpy helps us handle 
import numpy as np

#matplotlib helps us make basic plots
import matplotlib.pyplot as plt
import matplotlib.patches as patches
%matplotlib inline

#seaborn helps us make intricate plots
import seaborn as sns

#import google drive
from google.colab import drive

#give google colab access to your google drive
#click on the link in the output. It will take you to a new page where you will need to give permission to Google Colab to access your Google Drive.
#copy the link and paste it into the box below.
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
#view what folders you now have access to
!ls "/content/gdrive/MyDrive"

# Load Data

Note: You will get an error in the next cell if you did not save a shortcut of the "[CM_Datasets](https://drive.google.com/drive/u/0/folders/1WxSk_jD0x5rVSNiT8bM_58aMtG5Sqf4V)" folder to your drive. You *may* need to remount your drive to reflect new changes.

In [ ]:
# Load data

#provide the data location
filename = '/content/gdrive/MyDrive/CM_Datasets/water_potability.csv'

#ask pandas to grab the data from the location and name it as df (short for data frame)
df = pd.read_csv(filename)

#display dataframe
df

ph    Hardness  ...  Turbidity  Potability
0          NaN  204.890455  ...   2.963135           0
1     3.716080  129.422921  ...   4.500656           0
2     8.099124  224.236259  ...   3.055934           0
3     8.316766  214.373394  ...   4.628771           0
4     9.092223  181.101509  ...   4.075075           0
...        ...         ...  ...        ...         ...
3271  4.668102  193.681735  ...   4.435821           1
3272  7.808856  193.553212  ...   2.798243           1
3273  9.419510  175.762646  ...   3.298875           1
3274  5.126763  230.603758  ...   4.708658           1
3275  7.874671  195.102299  ...   2.309149           1

[3276 rows x 10 columns]

# Data Cleaning

Missing Values

In [ ]:
#search if there are any missing values in the data set
df.isnull().sum().sort_values()

Hardness             0
Solids               0
Chloramines          0
Conductivity         0
Organic_carbon       0
Turbidity            0
Potability           0
Trihalomethanes    162
ph                 491
Sulfate            781
dtype: int64

#**Question 2**: 
As you can see, there are a lot of missing values in the data. Given that there are 3276 patients, which columns should be dropped, and which columns should have their missing values replaced with the mean? (You can copy and paste from the table above).

Student Response: 

Drop: Sulfate

Replace: Trihalomethanes, pH

Unsure: 

If you are unsure about several columns, what makes you unsure about them?

Student Response: I am not sure if I should either drop pH or replace it with the mean, as it has less missing values than sulfate but many more missing values compared to trihalomethanes.

In [ ]:
#this code below will allow you to drop the columns you do not want.
#you need to put the column names in quotes
#uncomment the line of code below by deleting the '#' mark

df_drop = df.drop(['Sulfate'], axis=1)

In [ ]:
#fill in the remaining NaN with the means of each column
column_means = df_drop.mean()
df_drop = df_drop.fillna(column_means)

#**Question 3:**
In the code cell below, please display the number of NaNs in your clean data set. Were the NaN's replaced?

Yes.

In [ ]:
df_drop.isnull().sum().sort_values()

ph                 0
Hardness           0
Solids             0
Chloramines        0
Conductivity       0
Organic_carbon     0
Trihalomethanes    0
Turbidity          0
Potability         0
dtype: int64